File: meeting-scheduler/typescript/scheduler.ts
import { Room } from "./room";

export class Scheduler {
  private readonly rooms: Room[];
  constructor(rooms : Room[]){
    this.rooms = rooms;
  }

  public book(startTime: number, endTime: number): string {
    for(const room of this.rooms){
      const flag = room.book(startTime, endTime);
      if(flag){
        return room.name;
      }
    }
    throw new Error("All rooms are currently occupied")
  }

  //additional functionality to add a new room 
  public addRoom(room: Room): void {
    this.rooms.push(room)
  }
}

File: meeting-scheduler/go/scheduler.go
package main

import (
	"errors"
)


type Scheduler struct{
  Rooms []Room
}


func (scheduler *Scheduler) book(startTime int64, endTime int64) (string, error) {
  for i := range (scheduler.Rooms){
    room := &scheduler.Rooms[i];
    flag := room.book(startTime, endTime);
    if(flag == true){
      return room.name, nil;
    }
  }
  return "",errors.New("Cannot book room for this time")
}
